src/border_agent/border_agent.cpp

```diff
@@ -403,33 +377,44 @@ void BorderAgent::HandleMdnsState(Mdns::Publisher::State aState)
     return;
 }
 
-static uint64_t ConvertTimestampToUint64(const otTimestamp &aTimestamp)
+void BorderAgent::HandleBorderAgentMeshCoPServiceChanged(bool                        aIsActive,
+                                                         uint16_t                    aPort,
+                                                         const std::vector<uint8_t> &aOtMeshCoPTxtValues)
 {
-    // 64 bits Timestamp fields layout
-    //-----48 bits------//-----15 bits-----//-------1 bit-------//
-    //     Seconds      //      Ticks      //  Authoritative    //
-    return (aTimestamp.mSeconds << 16) | static_cast<uint64_t>(aTimestamp.mTicks << 1) |
-           static_cast<uint64_t>(aTimestamp.mAuthoritative);
-}
+    if (aIsActive != mBaIsActive || aPort != mMeshCoPUdpPort)
+    {
+        mBaIsActive     = aIsActive;
+        mMeshCoPUdpPort = aPort;
+    }
 
-#if OTBR_ENABLE_BORDER_ROUTING
-void AppendOmrTxtEntry(otInstance &aInstance, Mdns::Publisher::TxtList &aTxtList)
-{
-    otIp6Prefix       omrPrefix;
-    otRoutePreference preference;
+    mOtMeshCoPTxtValues.assign(aOtMeshCoPTxtValues.begin(), aOtMeshCoPTxtValues.end());
 
-    if (OT_ERROR_NONE == otBorderRoutingGetFavoredOmrPrefix(&aInstance, &omrPrefix, &preference))
+    // Parse extended address from the encoded data for the first time
+    if (!mIsInitialized)
     {
-        std::vector<uint8_t> omrData;
+        Mdns::Publisher::TxtList txtList;
+        otbrError                error =
+            Mdns::Publisher::DecodeTxtData(txtList, mOtMeshCoPTxtValues.data(), mOtMeshCoPTxtValues.size());
+
+        otbrLogResult(error, "Result of decoding MeshCoP TXT data from OT");
+        SuccessOrExit(error);
+
+        for (auto &entry : txtList)
+        {
+            if (entry.mKey == "xa")
```

Comment:
If I undrtansd this correctly, we are parsing the TXT data to extarct `"xa"` key so we can larn the ext address and then construct the "service instance name" from it?

How about if we directly ask the OT stack (e.g. using `otLinkGetExtendedAddress()`)? Would this work? I guess with this we wont need to parse TXT data anymore?

Reply:
Yes exactly. That sounds good to me. 

I'm trying to make RCP/NCP design transparent to the `BorderAgent` class and let this class work for both design. So I tried to not calling OT API here. My idea is that we make a GetExtendedAddress as an interface (asynchronized) and implement it for both RCP & NCP cases.  Then we call it in this class with a callback to receive ExtendedAddress as one of the condition to start the BorderAgent. Thoughts?

Reply:
Thanks. The current code (parsing TXT data) LGTM.